haskell-src/Concordium/MerkleProofs.hs

```diff
@@ -0,0 +1,566 @@
+parseBytes :: Int -> Parse r MerkleProof [Tag] ParseError BS.ByteString
+parseBytes len0 = Parse $ \ParserContext{..} ->
+    inner
+        (parserError . applyErrorContext parserReaderContext)
+        parserReaderContext
+        len0
+        parserInput
+        parserCont
+  where
+    inner _ ctx 0 inp cont = cont BS.empty inp ctx
+    inner parserError _ _ [] _ = parserError UnexpectedEndOfInput
+    inner parserError ctx len (RawData bytes : rest) cont = case compare (BS.length bytes) len of
+        LT -> inner parserError ctx (len - BS.length bytes) rest (cont . (bytes <>))
+        EQ -> cont bytes rest ctx
+        GT -> let (b0, b1) = BS.splitAt len bytes in cont b0 (RawData b1 : rest) ctx
+    inner parserError _ _ (SubProof _ : _) _ = parserError UnexpectedSubProof
+
+-- | Given a parser for a sub-proof, parse that sub-proof as part of a larger proof.
+--  The sub-proof parser should return (when successful) a builder for the byte string that should
+--  be hashed to give the root hash of the sub-proof.
+--  If a sub-proof is not next in the input, then the parser will take the next 32 bytes as a hash
+--  instead, treating the sub-proof as absent. If neither sub-proof nor 32 bytes are available, this
+--  will raise a 'ParseError'.
+--  The return value is the root hash of the subproof, and, if the sub-proof was parsed, the parsed
+--  result of the sub-proof.
+parseSubProof ::
+    Parse r MerkleProof [Tag] ParseError (a, Builder.Builder) ->
+    Parse r [MerkleBranch] [Tag] ParseError (SHA256.Hash, Maybe a)
+parseSubProof inside = Parse $ \pc@ParserContext{..} ->
+    let err = parserError . applyErrorContext parserReaderContext
+    in  case parserInput of
+            [] -> err UnexpectedEndOfInput
+            (RawData bs : rest)
+                | BS.null bs -> unParse (parseSubProof inside) pc{parserInput = rest}
+                | otherwise ->
+                    unParse
+                        ( do
+                            bytes <- parseBytes 32
+                            let hsh = SHA256.Hash (FBS.fromByteString bytes)
+                            return (hsh, Nothing)
+                        )
+                        pc
+            (SubProof sp : rest) ->
+                unParse
+                    inside
+                    pc
+                        { parserInput = sp,
+                          parserCont = \(subPT, builder) remaining ctx ->
+                            if isEmpty remaining
+                                then
+                                    let hsh = hashBuilder builder
+                                    in  parserCont
+                                            (hsh, Just subPT)
+                                            rest
+                                            parserReaderContext
+                                else parserError (applyErrorContext ctx ExpectedEndOfInput)
+                        }
+
+parseSize ::
+    (Num b) =>
```

Comment:
This is kind of too general. If you are parsing a U64 but give Word8 here the meaning here is not really as intended.

At least the preconditions must be documented.